### PR TITLE
fix: use int64 row offsets in persistent 2D copy

### LINKED ISSUE #201: copy_tensor: the persistent 2-D arm computes row offsets in int32, so a copy whose row extent passes 2 GiB reads and writes out of bounds
## Summary

`copy_tensor` has two 2-D kernels. `copy_2d_kernel` (`persistent=False`) casts the row index to int64 before scaling it by the row stride; `copy_2d_persistent_kernel` does not. Once `(M - 1) * stride(0)` reaches `2**31` the int32 product wraps negative, and the copy reads 2 GiB before the source and — when the destination is large too — writes 2 GiB before the destination. The uncast arm is the default: `copy_tensor(..., persistent=True)` is the signature default and five of the six in-tree call sites take it.

This needs either a big tensor (any 2-D tensor with more than `2**31` elements) or a large row stride; the 2-D branch has no contiguity assert, so a row slice of a large tensor also gets there.

https://github.com/ByteDance-Seed/Triton-distributed/blob/8260bc34398c2b8f36dc840fd22f741ca9294584/python/triton_dist/kernels/nvidia/memory_ops.py#L378-L383

## Reproduction

Both cases carve their tensors out of one allocation with a 2 GiB guard band in front, so the wrapped accesses land in memory the script owns and the result is a printed value rather than an illegal-access crash.

```python
import torch
from triton_dist.kernels.nvidia.memory_ops import copy_tensor

GiB, PAD = 1 << 30, 1 << 31
NUM_SMS = torch.cuda.get_device_properties(0).multi_processor_count

# case 1 -- 3 rows of 8 int8 taken 2**30 elements apart: row 2 is at offset 2**31.
buf = torch.empty(PAD + 2 * GiB + 8, dtype=torch.int8, device="cuda")
buf[:PAD].fill_(7)                                    # guard band, 2 GiB before src
src = torch.as_strided(buf, size=(3, 8), stride=(GiB, 1), storage_offset=PAD)
src.fill_(42)
for persistent in (True, False):
    dst = torch.full((3, 8), -1, dtype=torch.int8, device="cuda")
    copy_tensor(dst, src, num_sms=NUM_SMS, persistent=persistent)
    torch.cuda.synchronize()
    print(f"persistent={persistent!s:5s}  dst[2] = {dst[2].tolist()}")
del buf; torch.cuda.empty_cache()

# case 2 -- a plain contiguous 2.15 GB tensor: rows 262144.. wrap on BOTH the load and the store.
M, N = 262400, 8192
elems = M * N
buf = torch.empty(PAD + elems + PAD + elems, dtype=torch.int8, device="cuda")
buf[:PAD].fill_(7)
src = buf[PAD:PAD + elems].view(M, N); src.fill_(42)
for persistent in (True, False):
    buf[PAD + elems:PAD + elems + PAD].fill_(0)       # guard band, 2 GiB before dst
    dst = buf[PAD + elems + PAD:].view(M, N); dst.fill_(-1)
    copy_tensor(dst, src, num_sms=NUM_SMS, persistent=persistent)
    torch.cuda.synchronize()
    print(f"persistent={persistent!s:5s}  wrong rows {int((dst != 42).any(dim=1).sum())}/{M}"
          f"  bytes written into the guard band before dst: {int((buf[PAD + elems:PAD + elems + PAD] != 0).sum())}")
```

```
persistent=True   dst[2] = [7, 7, 7, 7, 7, 7, 7, 7]
persistent=False  dst[2] = [42, 42, 42, 42, 42, 42, 42, 42]
persistent=True   wrong rows 256/262400  bytes written into the guard band before dst: 2097152
persistent=False  wrong rows 0/262400  bytes written into the guard band before dst: 0
```

The `7`s in case 1 come from 2 GiB before the source. In case 2 the 256 wrong rows are exactly the rows from 262144 (`2**31 // 8192`) on, and 2097152 = 256 x 8192 is exactly those rows written to the wrong place.

## Environment

- Repo commit tested: `8260bc34398c2b8f36dc840fd22f741ca9294584` (main)
- GPU: NVIDIA B200 (sm_100), driver 595.71.05, CUDA 13.2
- torch 2.12.1+cu130, the repo's vendored Triton 3.4.0, Python 3.11

### PR DESCRIPTION
**Pull request by Levius-Fubuki**
## Purpose

`copy_2d_persistent_kernel` currently scales its row index by the source and
destination strides while the index is still `int32`. Once a valid row offset
reaches `2**31` elements, the product wraps and the kernel accesses memory
before the intended tensor.

The tilewise 2D copy already avoids this by widening its row index to
`tl.int64`. This PR applies the same address calculation to the persistent
path:

```text
before: int32 row index * row stride
after:  int64 row index * row stride
```

The widened row index is shared by the load and store address expressions.
The launch grid, block sizes, scheduling, and public `copy_tensor` API remain
unchanged.

Fixes #201.

### Regression test

```bash
PYTHONPATH=python python -m pytest -q \
  python/triton_dist/test/nvidia/test_memory_ops.py
```

The test constructs a `(3, 8)` `int8` view with row stride `2**30` and storage
offset `2**31`. Only 24 tensor elements are copied, while row 2 crosses the
signed 32-bit offset boundary.

It covers both directions:

- large-stride source: the pre-change persistent load reads the guard value
  `7` instead of the source value `42`;
- large-stride destination: the pre-change persistent store leaves row 2
  unchanged and overwrites the guard region;
- the tilewise implementation is used as a correctness control.

The test checks available device memory first and skips explicitly unless
approximately 5 GiB is free, including a 1 GiB safety margin.

### Lint and formatting

```bash
ruff check --config python/pyproject.toml \
  python/triton_dist/test/nvidia/test_memory_ops.py

yapf --diff \
  python/triton_dist/test/nvidia/test_memory_ops.py

python -m py_compile \
  python/triton_dist/test/nvidia/test_memory_ops.py

git diff --check
```

### RTX 4090 copy benchmark

Hardware/software: NVIDIA RTX 4090, driver 580.76.05, CUDA 13.0,
PyTorch 2.12.1+cu130. No GPU clock or kernel launch configuration changes.

Shape (FP16) | Old latency / bandwidth | New latency / bandwidth | Latency delta
--- | ---: | ---: | ---:
4096 x 4096 | 0.0850 ms / 789.59 GB/s | 0.0850 ms / 789.59 GB/s | 0.00%
8192 x 8192 | 0.3061 ms / 876.83 GB/s | 0.3062 ms / 876.74 GB/s | +0.03%
16384 x 8192 | 0.6021 ms / 891.65 GB/s | 0.6011 ms / 893.17 GB/s | -0.17%

Each result is the median of 200 timed iterations after 50 warmup iterations.
Effective bandwidth counts one read and one write. The observed differences
are within run-to-run noise. The standalone benchmark harness is kept outside
this correctness-only change so that the submitted diff remains focused.

## Test Result

Before the fix:

```text
persistent source row 2:      [7, 7, 7, 7, 7, 7, 7, 7]
tilewise source row 2:        [42, 42, 42, 42, 42, 42, 42, 42]
persistent destination row 2: [0, 0, 0, 0, 0, 0, 0, 0]
destination guard:            [42, 42, 42, 42, 42, 42, 42, 42]
```

After the fix:

```text
persistent source row 2:      [42, 42, 42, 42, 42, 42, 42, 42]
tilewise source row 2:        [42, 42, 42, 42, 42, 42, 42, 42]
persistent destination row 2: [42, 42, 42, 42, 42, 42, 42, 42]
destination guard:            [7, 7, 7, 7, 7, 7, 7, 7]
```

The focused RTX 4090 overflow harness, benchmark correctness checks,
`ruff-check`, `yapf --diff`, Python byte compilation, and
`git diff --check` passed locally.

## Limitations

The overflow regression intentionally uses a large sparse-stride allocation
to exercise real 64-bit GPU address arithmetic. It skips on devices without
enough free memory rather than replacing the runtime check with an
implementation-specific IR assertion.

The local GPU run used the released Triton-distributed compiler wheel to
compile the current `memory_ops.py` directly because the AutoDL environment
was not a full source build of the current repository. Full repository pytest
and cross-GPU CI remain for the upstream CI environment. The benchmark is an
isolated device-copy measurement, not an end-to-end model throughput claim.

## AI assistance

AI assistance was used during implementation, testing, benchmarking, and PR
drafting. The human contributor reviewed and takes responsibility for the
change and evidence.

### CODE REVIEW

**Comment by Levius-Fubuki** [comment]
Hi @KnowingNothing, since you authored the current memory ops implementation, could you please take a look when convenient? This fixes the persistent 2D copy row-offset overflow and adds regression coverage for both loads and stores. Thanks!

**Comment by KnowingNothing** [comment]
The fix looks good. Thank you!

**Review by KnowingNothing — APPROVED** [review]

**Comment by Levius-Fubuki** [comment]
Hi @KnowingNothing, the remaining AMD CI failure appears unrelated to this change. The `Build rocmshmem bind` step exits while checking out an Open MPI submodule with `fatal: reference is not a tree: 697a596dde68815fe50db3c2a75a42ddb41b5ef4`, before this PR is built or tested. The AMD MoRI SHMEM, CLA, and documentation checks passed. Could you please rerun the failed job or advise whether it can be treated as an infrastructure failure?
